File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort()
        m = len(potions)
        result = []

        for spell in spells:
            required_potion = (success + spell - 1) // spell 
            
            left, right = 0, m - 1
            while left <= right:
                mid = left + (right - left) // 2
                if potions[mid] < required_potion:
                    left = mid + 1
                else:
                    right = mid - 1
            
            count = m - left 
            result.append(count)
        
        return result

File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
       sort(potions.begin(), potions.end());
       int m = potions.size();
       vector<int> result;

       for (int spell : spells){
        long long required_potion = (success + spell - 1) / spell;

        int left = 0, right = m - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (potions[mid] < required_potion) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }  
        int count = m - left;
        result.push_back(count);
       }    
       return result;
    }
};